File: problem-solving/_lc-2351-easy-first-letter-to-appear-twice/1-solution.py
"""
https://leetcode.com/problems/first-letter-to-appear-twice/

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def repeatedCharacter(self, s: str) -> str:
        S = set()
        for c in s:
            if c in S:
                return c
            S.add(c)

File: problem-solving/_lc-28-easy-find-the-index-of-the-first-occurrence-in-a-string/1-solution.go
/*
https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

Complexity
----------
Time: O(m*n)
Space: O(m*n)

NOTE
----
This can be solved in linear time using standard pattern matching algorithms like
KMP, Rabin karp, ...
*/

func strStr(haystack string, needle string) int {
    var i, j int
    for i = 0; i < len(haystack) - len(needle) + 1; i++ {
        for j = 0; j < len(needle); j++ {
            if haystack[i + j] != needle[j] {
                break
            }
        }
        if j == len(needle) {
            return i
        }
    }
    return -1
}
